packages/beacon-node/src/chain/opPools/opPool.ts

```diff
@@ -369,19 +370,28 @@ export class OpPool {
   }
 
   /**
-   * Call after finalizing
-   * Prune blsToExecutionChanges for validators which have been set with withdrawal
-   * credentials
+   * Prune BLS to execution changes that have been applied to the state more than 1 block ago.
+   * In the worse case where head block is reorged, the same BlsToExecutionChange message can be re-added
+   * to opPool once gossipsub seen cache TTL passes.
    */
   private pruneBlsToExecutionChanges(
-    headState: CachedBeaconStateAllForks,
-    finalizedState: CachedBeaconStateAllForks | null
+    headBlock: allForks.SignedBeaconBlock,
+    headState: CachedBeaconStateAllForks
   ): void {
+    const {config} = headState;
+    const recentBlsToExecutionChanges =
+      config.getForkSeq(headBlock.message.slot) >= ForkSeq.capella
+        ? (headBlock as capella.SignedBeaconBlock).message.body.blsToExecutionChanges
+        : [];
+
+    const recentBlsToExecutionChangeIndexes = new Set(
+      recentBlsToExecutionChanges.map((blsToExecutionChange) => blsToExecutionChange.message.validatorIndex)
+    );
```

Comment:
This check looks insufficient. The prune handler is only called once per finalization event ~1 time / epoch. Then only the execution changes included in that block will be pruned, with all remaining changes staying there.

A more complete solution, using our state caches:

Attach new cache to the state:
- appliedBlsToExecutionChangesCurrentEpoch
- appliedBlsToExecutionChangesPreviousEpoch

During block processing you register the validator index of applied execution changes. During epoch transition you rotate the caches. Then to prune you use the `appliedBlsToExecutionChangesPreviousEpoch` list to prune those. This will give you the complete list of items to prune if there are no re-orgs deeper than 1 epoch, which is an okay compromise.

Reply:
> Then only the execution changes included in that block will be pruned, with all remaining changes staying there.

just to clarify, execution changes in the head block is __not__ pruned, others will be pruned. So we store BLSToExecutionChange messages in max 33 slots

> A more complete solution, using our state caches

the solution works, however I think it's too much for this feature since we have to apply ImmutableJS to maintain these 2 arrays like in EIP-6110 implementation #6042 Also it's tricky in the edge case

I think there are pros and cons for these 2 solutions:

|#|BLSToExecutionChange messages in block slot in opPool|pros|cons|
|-|-----|----|----|
|current| 3 epochs|straightforward|have to query the finalized state|
|no state cache (this PR)| 1 slot to 33 slots|no state cache is needed, lighthouse runs it since capella|cover 1-slot reorg scenario only|
|state cache (Lion's suggestion)|32 slots to 64 slots|cover up to 1-epoch reorg scenario|need to have 2 ImmutableJS arrays maintained, also need to think about edge case when finalized checkpoint happens >4s, after clock epoch|

I prefer the no state cache approach as it's worked for lighthouse since capella and in the worse case if the reorg happen, it'll not affect block reward (if it's the bn's turn to propose block) and messages will be added back once gossip `seenTTL` passes anyway, I'd like to know others' opinions cc @wemeetagain @g11tech

Reply:
I think what @tuyennhv proposed here is sufficient and his reasoning is sound.

Reply:
Sounds good, thanks for the table! please monitor metrics after deploying to check the pool is not growing